18.md

```diff
@@ -0,0 +1,153 @@
+  uri = "http://localhost:3000/test.csv"
+  uri_double = URI.parse(uri)
+  expect(URI).to receive(:parse).with(uri).and_return(uri_double)
+
+  file_contents = File.read("fixtures/test.csv")
+  allow(uri_double).to receive(:read).and_return(file_contents)
+  csv_double = CSV.parse(file_contents, headers: true)
+  expect(CSV).to receive(:parse).with(file_contents, headers: true).and_return(csv_double)
+  expect(FileInfo.new(uri).column_count).to eq(csv_double.first.headers.count)
+end
+```
+
+That's not a unit test, it's a full duplicate of our implementation.
+
+A couple of newsletters back I [suggested a way to visualize why this code is hard to test](http://tinyletter.com/ianwhitney/letters/code-smells-and-integration-tests). If you inline all the method calls used in 'column_count' you get:
+
+```
+CSV.parse(URI.parse(uri).read, headers: true).first.headers.count
+```
+
+Which is a lot of work for one method. And nearly all of that work is creating objects. What happens if we extract that work and only give `column_count` just the object it needs? First, let's extract the URI creation.
+
+```ruby
+class FileInfo
+  def new(uri)
+    self.uri = uri
+    get_file
+  end
+
+  #...
+  attr_accessor :uri
+  attr_reader :file
+
+  def get_file
+    @file = CSV.parse(uri.read, headers: true)
+  end
+end
+
+uri = URI.parse(http://localhost:3000)
+FileInfo.new(uri).column_count
+```
+
+That reduces our unit test to:
+
+```ruby
+describe "column_count" do
+  uri_double = URI.parse("http://localhost:3000/test.csv")
+
+  file_contents = File.read("fixtures/test.csv")
+  allow(uri_double).to receive(:read).and_return(file_contents)
+  csv_double = CSV.parse(file_contents, headers: true)
+  expect(CSV).to receive(:parse).with(file_contents, headers: true).and_return(csv_double)
+  expect(FileInfo.new(uri).column_count).to eq(csv_double.first.headers.count)
+end
+```
+
+Better. Still a ways to go. Let's extract out CSV.
+
+```ruby
+class FileInfo
+  def new(csv)
+    file = self.csv
+  end
+
+  #...
+  attr_accessor :file
+end
+
+uri = URI.parse(http://localhost:3000)
+file = CSV.parse(uri.read, headers: true)
+FileInfo.new(file).column_count
+```
+
+Which brings our unit test to:
+
+```ruby
+describe "column_count" do
+  csv_double = CSV.parse(File.read("fixtures/test.csv"), headers: true)
+  expect(FileInfo.new(csv_double).column_count).to eq(csv_double.first.headers.count)
+end
+```
+
+Which fairly succinctly tests the behavior that the method provides.
+
+Obviously I'm cheating a little bit here. I extracted the URI and CSV code out of the class but I didn't put it anywhere. It's just hanging out in a nameless void of Ruby. In the real world you'd put that code in their own classes. `FileInfo` expects a header-parsed CSV file, so make a class that provides that
+
+```
+class HeaderCSV < SimpleDelegator
+  def self.build(file)
+    new(CSV.parse(file.read, headers: true))
+  end
+end
+
+file = HeaderCSV.new(URI.parse('http://localhost:3000').read)
```

Comment:
Reader Tim has pointed out my mistake here. Thanks Tim!

It should read

`file = HeaderCSV.build(URI.parse('http://localhost:3000').read)`